### PR TITLE
Json

### PR DESCRIPTION
**Pull request by jzkelter**
This implements a table:from-json primitive and also includes tests for that new primitive. Right now, table:from-json can only read in a json object, not a json list, but if someone wants a list of json objects read it, it is easy to just wrap that in a json object.

### CODE REVIEW

File: src/TableExtension.java [hunk 1]
```diff
@@ -17,8 +17,22 @@
...
+import java.io.File;
+import java.io.FileReader;
+import com.google.gson.Gson;
+import com.google.gson.internal.LinkedTreeMap;
+
+// import org.json.JSONObject;
+// import org.json.JSONTokener;
+// import org.json.JSONParser;
+// import com.typesafe.config.ConfigFactory;
+// import com.typesafe.config.Config;
```
Comment (LaCuneta) [inline src/TableExtension.java:34]: Can remove these unused commented imports.

File: build.sbt [hunk 2]
```diff
@@ -14,4 +14,11 @@ javaSource in Compile := baseDirectory.value / "src"
 javacOptions ++= Seq("-g", "-deprecation", "-Xlint:all", "-Xlint:-serial", "-Xlint:-path",
   "-encoding", "us-ascii")
 
+libraryDependencies ++= Seq(
+   "com.google.code.gson" % "gson" % "2.8.6"
+)
+    // "org.json" % "json" % "20201115")
```
Comment (LaCuneta) [inline build.sbt:20]: Can remove this unused, commented dependency.

File: src/TableExtension.java [hunk 3]
```diff
@@ -92,6 +107,31 @@ public Table(long id) {
...
+      tables.put(this, next);
+      id = next;
+      next++;
+
+      for (java.util.Map.Entry<Object,Object> entry : map.entrySet()) {
+        this.put(entry.getKey(), getTableValue(entry.getValue()));
+      }
+    }
+
+    private Object getTableValue(Object getValue) {
```
Comment (LaCuneta) [inline src/TableExtension.java:120]: The `get` part of `getValue` makes me think it's supposed to be a function, we usually try not to use verbs for things that are just data.  Maybe just `value` would work?

File: src/TableExtension.java [hunk 4]
```diff
@@ -343,6 +383,43 @@ public Object report(Argument args[], Context context)
...
+    public Object report(Argument args[], Context context)
+        throws ExtensionException, LogoException {
+
+      FileManager fm = ((ExtensionContext) context).workspace().fileManager();
+
+      try {
+        String path = fm.attachPrefix(args[0].getString());
+        File file = new File(path.toString());
+        if (!file.exists()) {
+          throw new ExtensionException (args[0].get().toString() + " does not exist.");
```
Comment (LaCuneta) [inline src/TableExtension.java:406]: Prefer no space between a constructor and its open paren, so `ExtensionException(`

File: src/TableExtension.java [hunk 5]
```diff
@@ -343,6 +383,43 @@ public Object report(Argument args[], Context context)
...
+        }
+        Gson gson = new Gson();
+
+        try {
+          java.util.Map<Object,Object> map = gson.fromJson(new FileReader(path), java.util.Map.class);
+          return new Table(map);
+        } catch (com.google.gson.JsonSyntaxException e) {
+          throw new ExtensionException ("Error trying to read the JSON file. It is probably missing a colon or comma. See the line number on the next line: " + e.getMessage());
```
Comment (LaCuneta) [inline src/TableExtension.java:414]: Prefer no space between a constructor and its open paren, so `ExtensionException(`

**Review by LaCuneta — CHANGES_REQUESTED** [review]
Thanks for your work on this!  It's great to have the tests there, and especially the negative test.

My usual proviso when reviewing PRs on NetLogo's older extensions:  You might find examples of the extension doing things differently than I ask for in my comments/suggestions here.  Our preferred coding style and standards have changed, so as code is changed or added I try to make sure we're adhering to the latest for the new work.

When I did a `extension table` from a NetLogo sbt console, I got some warnings for unchecked conversions/casts.  These obviously don't stop things from working, but we prefer to avoid warnings during compilation/testing/packaging as much as we can.  Most of the warnings here looked like auto-conversions from `Map` to `Map<Object, Object>`.  I think the best way to resolve these is to use `Map<?, ?>` with the type wildcard `?` instead of `Object` wherever you need to fix the type of an argument, variable, or cast.  That will tell the Java compiler you really don't know the type parameters for the `Map`, and you also don't really care so it'll stop bugging you.  You may have to do the same with `Map.Entry<Object, Object>` and the `ArrayList` cast to clear all the warnings.

I notice a lot of use of fully-qualified names, like `java.util.Map`, `java.io.IOException`, `com.google.gson.JsonSyntaxException`, and probably a few more.  We prefer adding imports at the top of the file for the used classes, and then referencing them by the short name, so `Map`, `IOException`, and `JsonSyntaxException`.

When I ran `te table` from a NetLogo sbt console, I got a different error for the `TableFromJSON` language test, saying `extensions/table/examples/sample-malformed-json.json does not exist`.  I think this is because at one point `examples/` was added to the `.gitignore` file, so when the malformed JSON file was added, it wasn't picked up for your commit.  I think the `examples/` folder should be removed from the ignore file, and the malformed file could then be added for the test.

I noticed that the `gson-2.8.6.jar` is included in these commits.  We do not generally store binary files as part of our code repositories.  You can add `*.jar` to the `.gitignore` file to make sure it doesn't get added again (instead of just `table.jar`).

How would you feel about renaming the prim `from-json-file`?  I'm thinking ahead to if we add this functionality to the NetLogo Web version, we won't be able to match the local file loading behavior directly.  In both this and NLW we could add `from-json-string` prim that loads JSON from a string and couple that with the `fetch` extension to get the data in NLW (and desktop, too).  The `from-json-file` would match how to the CSV extension splits things up (`from-file` and `from-string`), and could allow us to have that same flexibility of data source for NLW.

I also added some smaller style comments to the code directly.  Let me know if you have any questions.

**Comment by jzkelter** [comment]
I think I've dealt with all the issues you raised. However, I didn't always get warnings when compiling even at the beginning (but sometimes do). So, it is possible there are still a few changes to make to get rid of warnings.

**Review by LaCuneta — CHANGES_REQUESTED** [review]
I did get a couple more warnings, line 128, `ArrayList` cast should be `ArrayList<?>` and line 413, the type for the `map` variable should be `Map<?, ?>`.  For future reference, if you do an `sbt clean` in the Table repository it should reset and give you the full compilation and warnings info when you next do an `sbt compile`.  

Otherwise, these changes all look good.  Tested it out and things worked as expected.  Once the above couple types are fixed, we can merge.

**Comment by jzkelter** [comment]
Okay I made those changes. When I compile the way you said in the last comment, I got those warnings. I was compiling in the NetLogo directory by first doing `./sbt` to enter the sbt console and then `extension time` and wasn't getting warnings.

**Comment by LaCuneta** [comment]
Hey Jacob, awesome, thanks!  I ran the package, no warnings or errors.  I ran the tests and everything is good.  

I'm super sorry to ask for one more thing.  I should've noticed on my first review, but I missed it: there was no documentation added for the new prim.  All you need to do to update is change the `documentation.conf` to include it (see the other prims in there as a reference), then in `sbt` in this package, run `extensionDocument` and that will generate a new `README.md` file.  You can then check both of those in in one more commit and this will (really, for real!) be ready to go.  Let me know if you have questions, and again, my apologies for missing it.

**Comment by jzkelter** [comment]
I added documentation, let me know if it is fine.

**Review by LaCuneta — CHANGES_REQUESTED** [review]
This looks good to me, thank you again!

**Review by LaCuneta — APPROVED** [review]

**Pull request merged**